File: FastMines_swing/src/main/java/fmg/swing/utils/Timer.java
package fmg.swing.utils;

import java.awt.event.ActionListener;
import java.util.Date;
import java.util.function.Consumer;

import fmg.common.ui.ITimer;

/** Swing timer. Worked in UI thread */
public class Timer implements ITimer {

    private javax.swing.Timer swTimer;
    private long interval = 200;
    private Consumer<ITimer> callback;
    private boolean paused = true;
    /** if played: it`s start time; if paused: time offset */
    private long started;

    @Override
    public long getInterval() { return interval; }

    @Override
    public void setInterval(long delay) {
        this.interval = delay;
        if (swTimer != null)
            swTimer.setDelay((int)delay);
    }

    private void myCallback() {
        if (callback != null)
            callback.accept(this);
    }

    @Override
    public void setCallback(Consumer<ITimer> callback) {
        this.callback = callback;
    }

    @Override
    public void start() {
        if (!paused)
            return;

        paused = false;
        started = new Date().getTime() - started; // apply of pause delta time

        if (swTimer == null) {
            swTimer = new javax.swing.Timer((int)interval, ev -> myCallback());
            swTimer.setRepeats(true);
        }
        swTimer.start();
    }

    @Override
    public void pause() {
        if (paused)
            return;

        paused = true;
        started = new Date().getTime() - started; // set of pause delta time
        if (swTimer != null)
            swTimer.stop();
    }

    @Override
    public void reset() {
        pause();
        started = 0;
    }

    @Override
    public long getTime() {
        if (paused)
            return started;
        return new Date().getTime() - started;
    }

    @Override
    public void setTime(long time) {
        if (paused)
            started = time;
        else
            started = new Date().getTime() - time;
    }

    @Override
    public void close() {
        callback = null;
        if (swTimer != null) {
            swTimer.stop();
            for (ActionListener al : swTimer.getActionListeners())
                swTimer.removeActionListener(al);
            swTimer = null;
        }
    }

}

File: FastMines_Ava/src/fmg/ava/utils/Timer.cs
using System;
using Avalonia.Threading;
using Fmg.Common.UI;

namespace Fmg.Ava.Utils {

    public class Timer : ITimer {

        private DispatcherTimer _timer;
        private long _interval = 200;
        private Action<ITimer> _callback;
        private bool _paused = true;
        /// <summary> if played: it`s start time; if paused: time offset </summary>
        private long _started;

        /// <summary> in miliseconds </summary>
        public long Interval {
            get { return _interval; }
            set {
                _interval = value;
                if (_timer != null)
                    _timer.Interval = TimeSpan.FromMilliseconds(value);
            }
        }

        private void OnTick(object sender, object e) {
            _callback?.Invoke(this);
        }

        public Action<ITimer> Callback { set { _callback = value; } }

        public void Start()
        {
            if (!_paused)
                return;

            _paused = false;
            _started = DateTimeOffset.Now.ToUnixTimeMilliseconds() - _started; // apply of pause delta time

            if (_timer == null) {
                _timer = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(Interval) };
                _timer.Tick += OnTick;
            }
            _timer.Start();
        }

        public void Pause() {
            if (_paused)
                return;

            _paused = true;
            _started = DateTimeOffset.Now.ToUnixTimeMilliseconds() - _started; // set of pause delta time
            if (_timer != null)
                _timer.Stop();
        }

        public void Reset() {
            Pause();
            _started = 0;
        }

        public long Time {
            get {
                if (_paused)
                    return _started;
                return DateTimeOffset.Now.ToUnixTimeMilliseconds() - _started;
            }
            set {
                if (_paused)
                    _started = value;
                else
                    _started = DateTimeOffset.Now.ToUnixTimeMilliseconds() - value;
            }
        }

        private void Clean() {
            if (_timer == null)
                return;

            _timer.Tick -= OnTick;
            _timer.Stop();
            _timer = null;
            _callback = null;
        }

        public void Dispose() {
            Clean();
        }

    }

}
